sklearn/tree/tests/test_tree.py

```diff
@@ -1233,8 +1232,7 @@ def test_max_leaf_nodes_max_depth():
     k = 4
     for name, TreeEstimator in ALL_TREES.items():
         est = TreeEstimator(max_depth=1, max_leaf_nodes=k).fit(X, y)
-        tree = est.tree_
-        assert_greater(tree.max_depth, 1)
+        assert_greater(est.get_max_depth(), 1)
```

Comment:
I don't get this test. Why should it be greater than ``max_depth``?

Reply:
lol, don't hate me, I just change _a_ test to use the new functions. But I suppose that test is valid because `X` and `y` are manually designed in a way that we know the tree will be deeper if allowed, and this makes sure `max_depth` is ignored by the estimator if `max_leaf_nodes` is given. At least that's what the comment says.

Reply:
ohh... I didn't read the comment. That behavior is... unexpected to me? Should we raise a warning in that case? Do we?

Reply:
Also, there is no test that shows that the depth can be smaller than max_depth, right? We should have some slightly more explicit test, I think.

Reply:
@amueller, this actually seems a bug to me, which I think can be fixed [this way](https://github.com/scikit-learn/scikit-learn/compare/master...adrinjalali:bug/tree/maxdepthandleafnodes).

On master/this branch, we have the following:

```
>>> from sklearn.datasets import load_iris
>>> from sklearn.tree import DecisionTreeClassifier
>>> X, y = load_iris(return_X_y=True)
>>> 
>>> clf = DecisionTreeClassifier(random_state=0, max_depth=1, max_leaf_nodes=100)
>>> clf = clf.fit(X, y)
>>> clf.get_max_depth()
2
>>> clf.get_n_leaves()
3
>>> 
>>> clf = DecisionTreeClassifier(random_state=0, max_depth=1)
>>> clf = clf.fit(X, y)
>>> clf.get_max_depth()
1
>>> clf.get_n_leaves()
2
>>> 
>>> clf = DecisionTreeClassifier(random_state=0, max_leaf_nodes=100)
>>> clf = clf.fit(X, y)
>>> clf.get_max_depth()
5
>>> clf.get_n_leaves()
9
```
Which shows that the `max_depth` is not actually ignored, it's just an overshoot by 1. And the issue is here: https://github.com/scikit-learn/scikit-learn/blob/3804ccd2770ac4c026a823d019091917e4a2c70e/sklearn/tree/_tree.pyx#L453

After the proposed fix, we'd have:
```
>>> from sklearn.datasets import load_iris
>>> from sklearn.tree import DecisionTreeClassifier
>>> X, y = load_iris(return_X_y=True)
>>> 
>>> clf = DecisionTreeClassifier(random_state=0, max_depth=1, max_leaf_nodes=100)
>>> clf = clf.fit(X, y)
>>> clf.get_max_depth()
1
>>> clf.get_n_leaves()
2
>>> 
>>> clf = DecisionTreeClassifier(random_state=0, max_depth=1)
>>> clf = clf.fit(X, y)
>>> clf.get_max_depth()
1
>>> clf.get_n_leaves()
2
>>> 
>>> clf = DecisionTreeClassifier(random_state=0, max_leaf_nodes=100)
>>> clf = clf.fit(X, y)
>>> clf.get_max_depth()
5
>>> clf.get_n_leaves()
9
```
which then makes sense. I think that change needs a `Bug Fix` entry in whats_new, since it changes the behavior? And should I include it in this PR or create a new one?